flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.shuffle;
+
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.runtime.deployment.InputGateDeploymentDescriptor;
+import org.apache.flink.runtime.deployment.ResultPartitionDeploymentDescriptor;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.apache.flink.runtime.executiongraph.PartitionInfo;
+import org.apache.flink.runtime.io.network.api.writer.ResultPartitionWriter;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.partition.PartitionProducerStateProvider;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.taskexecutor.TaskExecutor;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/**
+ * Interface for the implementation of shuffle service environment locally on task executor.
+ *
+ * <p>Input/Output interface of local shuffle service environment is based on memory {@link Buffer}s.
+ * The task can write shuffle data into the buffers,
+ * obtained from the created here {@link ResultPartitionWriter}s
+ * and read the buffers from the created here {@link InputGate}s.
+ *
+ * <h2>Lifecycle management.</h2>
+ *
+ * <p>The interface contains method's to manage the lifecycle of the local shuffle service environment:
+ * <ol>
+ *     <li>{@code start} is called when the {@link TaskExecutor} is being started.</li>
```

Comment:
Not sure whether I would include the `TaskExecutor` here. It is a concern of the `TaskExecutor` when this service is started and should not matter to the service implementation. The important bit is that `start` is being called before this service is used.